### PR TITLE
Add async validation support

### PR DESCRIPTION
**Pull request by imbrn**
Related to #25

### CODE REVIEW

File: src/v8n.js [hunk 1]
```diff
@@ -221,9 +263,75 @@ const core = {
...
 
+function executeAsyncRules(value, rules) {
+  return new Promise((resolve, reject) => {
+    executeAsyncRulesAux(value, rules, 0, resolve, reject);
+  });
+}
+
+function executeAsyncRulesAux(value, rules, index, resolve, reject) {
+  if (index < rules.length) {
+    const rule = rules[index];
```
Comment (sebastianbarfurth) [inline src/v8n.js]: Use ``array.shift()`` here to save index passing.

File: src/v8n.js [hunk 2]
```diff
@@ -221,9 +263,75 @@ const core = {
...
+    const result = Promise.resolve(rule.fn(value));
+
+    result.then(valid => {
+      if (valid !== rule.invert) {
+        executeAsyncRulesAux(value, rules, index + 1, resolve, reject);
```
Comment (sebastianbarfurth) [inline src/v8n.js]: Pass shifted rules here instead of the entire array and the index.

File: src/v8n.js [hunk 3]
```diff
@@ -221,9 +263,75 @@ const core = {
...
+      } else {
+        reject(rule);
+      }
+    });
+  } else {
+    resolve();
```
Comment (sebastianbarfurth) [inline src/v8n.js]: Possibly resolve to the value as to give something useful?

File: src/v8n.js (hunk 2 above)
Comment (imbrn) [reply src/v8n.js]: Nice, that's a good improvement for clean code.

**Comment by olstenlarck** [comment]
I'm more for `testAsync` and `checkAsync`.

**Comment by sebastianbarfurth** [comment]
I would agree with @olstenlarck - it makes more sense in my opinion.

**Comment by imbrn** [comment]
On then, so we're going to use `*Async` methods. Thank you all.

File: src/v8n.js [hunk 4]
```diff
@@ -290,19 +290,21 @@ const core = {
    * Performs asynchronous failed-rule based validation.
    *
    * When this function is used it performs the validation process
-   * asynchronously, and it returns a promise that resolves to nothing when the
-   * validated value is valid and rejects with the first failed
-   * {@link Rule rule} when it's invalid.
+   * asynchronously, and it returns a promise that resolves to the validated
+   * value when it's valid and rejects with a {@link ValidationException} when
+   * it's invalid or when an exception occur.
```
Comment (sebastianbarfurth) [inline src/v8n.js]: **occurs** for occur

File: src/v8n.js [hunk 5]
```diff
@@ -290,19 +290,21 @@ const core = {
...
    * > [Validation](#Validation) section.
    *
    * > For a validation strategy with non promise-based rules, you'd better use
    * > the [check](#check) function.
    *
+   * @see ValidationException
    * @param {any} value the value to be validated
-   * @returns {Promise} promise that rejects with the first-failed
-   * {@link Rule rule} when value is invalid
+   * @returns {Promise} promise that rejects with a {@link ValidationException}
```
Comment (sebastianbarfurth) [inline src/v8n.js]: I don't think this paragraph makes sense?

**Comment by sebastianbarfurth** [comment]
I would like to add that I feel ``testAsync()`` should reject and not resolve if the validation fails. That's exactly what reject and resolve are for. Opinion?

**Comment by imbrn** [comment]
That's exactly what we get with the `checkAsync` function. So what can we do in order to make them different?

**Comment by sebastianbarfurth** [comment]
I feel like both aren't necessary for async since one of the points of Promises is that they either reject or resolve and you can work based off of that with ``.then()`` and ``.catch()``. If we implement them and then we resolve to errors we are breaking convention.

Can you think of scenarios where both would make sense? Like where you would need to have a differentiated check and test for async validation.

I can see it for non-async, since there you might pass ``test()`` directly to an `if` while you might use ``check()`` inside a `try-catch`. The same doesn't apply to async.

**Comment by imbrn** [comment]
You're right. When I decided on these two functions, I was thinking about their names only. But I think we have to keep only the `testAsync`.

**Comment by sebastianbarfurth** [comment]
Yeah, but I feel the functionality should be that of ``checkAsync`` and the name should be ``testAsync``. Of course we might also just call it ``async()``?

**Comment by imbrn** [comment]
Yes, we'll use the `checkAsync` functionality but with the `testAsync` name. I don't think we should use just `async`, it sounds a little confusing to me.

**Comment by sebastianbarfurth** [comment]
Fair enough. Sounds good though. The only useful variation I could imagine might be ``testAsyncAll()`` at some point for a resulting exception contains all rules that failed instead of just the first one.

**Comment by imbrn** [comment]
I see. Maybe it could be `testAllAsync`.

**Comment by sebastianbarfurth** [comment]
Yeah, sounds better and more consistent. Not necessary for the moment though.

By the way I believe this pull-request merged would constitute a SemVer minor update to 1.1.0 since it's completely new functionality in a backward compatible manner.

**Comment by imbrn** [comment]
It closes issue #25

**Pull request merged**